owlkettle/autoform.nim

```diff
@@ -0,0 +1,135 @@
+import std/[options, json, times, macros, strutils, sequtils]
+import ./dataentries
+import ./adw
+import ./widgetutils
+import ./guidsl
+import ./widgetdef
+import ./widgets
+
+macro getField*(someType: untyped, fieldName: static string): untyped =
+  nnkDotExpr.newTree(someType, ident(fieldName))
+
+proc `%`*(t: tuple): JsonNode =
+  result = newJObject()
+  for field, value in t.fieldPairs:
+    result.add(field, %value)
+
+## General Fields
+
+proc toFormField(state: auto, fieldName: static string, typ: typedesc[SomeFloat]): Widget =
+  return gui:
+    ActionRow:
+      title = fieldName
+      NumberEntry() {.addSuffix.}:
+        value = state.getField(fieldName)
+        xAlign = 1.0
+        maxWidth = 8
+        proc changed(value: float) =
+          state.getField(fieldName) = value
+
+proc toFormField(state: auto, fieldName: static string, typ: typedesc[SomeInteger]): Widget =
+  return gui:
+    ActionRow:
+      title = fieldName
+      NumberEntry() {.addSuffix.}:
+        value = state.getField(fieldName).float
+        xAlign = 1.0
+        maxWidth = 8
+        proc changed(value: float) =
+          state.getField(fieldName) = value.int
+
+proc toFormField(state: auto, fieldName: static string, typ: typedesc[string]): Widget =
+  return gui:
+    ActionRow:
+      title = fieldName
+      Entry() {.addSuffix.}:
+        text = state.getField(fieldName)
+        proc changed(text: string) =
+          state.getField(fieldName) = text
+
+proc toFormField(state: auto, fieldName: static string, typ: typedesc[bool]): Widget =
+    return gui:
+      ActionRow:
+        title = fieldName
+        Box() {.addSuffix.}:
+          Switch() {.vAlign: AlignCenter, expand: false.}:
+            state = state.getField(fieldName)
+            proc changed(newVal: bool) =
+              state.getField(fieldName) = newVal
+
+proc toFormField(state: auto, fieldName: static string, typ: typedesc[object | ref object | tuple | seq]): Widget =
```

Comment:
> I've decided to just move over to the approach where you defined a custom toFormField for every custom type you have.
> Generating forms for every normal type in an action Row is way too annoying.

This should also be removed then (?)

Reply:
Not quite, I do need a default on what should happen for those types.
That is simply that I turn them into json and the user can update that JSON string (And those updates remain as long as the string can still be parsed into the custom type.

That way I don't *force* the user to define a `toFormField` proc for their type, it is merely an option.

If an (ugly) default like this is not desired I'll throw it out of course, was just my default approach.

Reply:
The issue I have with JSON & Nim is that the mapping between the two is not bijective. There are Nim objects that cannot be mapped to JSON (e.g. cyclic references) and different types might map to the same JSON. (I am actually not exactly sure what happes for recursive types, though I would assume that it just creates an infinite recursion (?)).

From a UI perspective, the editing experience of typing JSON into an entry is also not really desirable.

There is however still the question of what to do with what to do with types that do not have a `toFormField` implementation. I think that since you want to encourage creating custom controls for your types (an appraoch I strongly agree with), it might be fine to just show some icon with a tooltip that says something like "Unable to edit [TypeName], please override toFormField.".

Reply:
I like the "Add a dummy-widget"-approach! I'll do that one